src/python/espressomd/zn.py

```diff
@@ -0,0 +1,436 @@
+        return radius_list
+
+    def get_bonds(self):
+        bonds = []
+        for p in self.particles:
+            if not p.bonds:
+                continue
+            for bond in p.bonds:
+                partner_id = bond[-1]
+                dist = np.linalg.norm(
+                    self.positions[p.id] - self.positions[partner_id])
+
+                # Remove bonds that are too long due to pbc
+                if dist > min(self.system.box_l) / 2: 
+                    continue
+                bonds.append([p.id, partner_id, 1])
+        return bonds
+
+
+znjson.config.register(EspressoConverter)
+
+
+class Visualizer():
+    def __init__(self,
+                 system: System = None,
+                 port: int = 1234,
+                 token: str = secrets.token_hex(4),
+                 folded: bool = True,
+                 colors: dict = None,
+                 radii: dict = None,
+                 bonds: bool = False,
+                 jupyter: bool = True,
+                 vector_field=None,
+                 ):
+
+        self.system = system
+        self.params = {
+            "folded": folded,
+            "colors": colors,
+            "radii": radii,
+            "bonds": bonds,
+            "vector_field": vector_field
+        }
+
+        self.url = "http://127.0.0.1"
+        self.frame_count = 0
+        self.port = port 
+        self.token = token
+
+        # The server is started in a subprocess, and we have to wait for it to start
+        print("Starting ZnDraw server, this may take a few seconds")
+        self._start_server()
+        time.sleep(10)
+        self._start_zndraw()
+        time.sleep(1)
+
+        if vector_field is not None:
+            if vector_field.arrow_config is None:
+                # Default arrow configuration
+                self.arrow_config = {'colormap': [[-0.5, 0.9, 0.5], [-0.0, 0.9, 0.5]],
+                                     'normalize': True,
+                                     'colorrange': [0, 1],
+                                     'scale_vector_thickness': True,
+                                     'opacity': 1.0}
+            else:
+                self.arrow_config = vector_field.arrow_config
+
+        if jupyter:
+            self._show_jupyter()
+        else:
+            # Problems with the server being terminated at the end of the script.
+            raise NotImplementedError("Only Jupyter is supported for now")
+            # webbrowser.open_new_tab(self.address)
+
+    def _start_server(self):
+        """
+        Start the ZnDraw server through a subprocess
+        """
+        self.socket_port = get_port(default=6374)
+
+        self.server = subprocess.Popen(["zndraw", "--no-browser", f"--port={self.port}", f"--storage-port={self.socket_port}"],
+                                       stdout=subprocess.DEVNULL,
+                                       stderr=subprocess.DEVNULL
+                                       )
```

Comment:
I am not aware of a better alternative than `subprocess.Popen()`, because both `subprocess.run` and `subprocess.check_call` are blocking afaik. The server has to be permanently running in the background and the command used to starting the server does not complete until the server is shutdown, thus we need a non-blocking subprocess call. I'm aware `Popen` can be dangerous, but the only alternative Im aware of is starting the server manually in a commandline and then connecting to the server using jupyter in another commandline. If you have any suggestions on a safer option, I would be happy to use it.

Reply:
Good point. `Popen()` is our only option for ansynchronous processes.